woocommerce/woocommerce
# Add a retry on the setup test environment step

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

There have been a number of CI failures recently (particularly in our trunk merge runs) that are the result of Docker just failing to spin up the environment properly (or failing to do it in time).  These look like a failure in the "Start Test Environment" step, where it has trouble connecting to the database ([for example](https://github.com/woocommerce/woocommerce/actions/runs/11147432123/job/30981781193))

```
✖ Error while running docker compose command.
 Container 36f129a9d109[27](https://github.com/woocommerce/woocommerce/actions/runs/11147432123/job/30981781193#step:5:28)4d16f01b89c88bf54c-tests-mysql-1  Running
 Container 36f129a9d109274d16f01b89c88bf54c-tests-wordpress-1  Running
Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `tests-mysql` could not be established. This could mean your host’s database server is down.
 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed with exit code 1.
/home/runner/work/woocommerce/woocommerce/plugins/woocommerce:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @woocommerce/plugin-woocommerce@9.5.0 env:test: `pnpm env:dev`
Exit status 1
```

There's likely an issue with Docker, the mysql docker image, or `wp-env` that is causing this, but it's so intermittent, it's hard to track down what might be the root cause.

This PR introduces a retry on the "Start Test Environment" step.  It will retry the step a maximum of 2 times before it fails.  That should mostly take care of the issue, as this failure frequently passes when the failure is retried.

For reference, [this is what this PR is using](https://github.com/nick-fields/retry)



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure that CI passes on this PR.
2. It's hard to trigger the Docker failure, but [here's a CI run](https://github.com/woocommerce/woocommerce/actions/runs/11149016960/job/30986970514?pr=51871) to ensure that the Setup Environment Step is still working.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 
Update to a GH workflow step to retry it.

</details>
